src/Native/Unix/System.Native/pal_signal.cpp

```diff
@@ -0,0 +1,286 @@
+        }
+    }
+}
+
+static void CloseSignalHandlingPipe()
+{
+    assert(g_signalPipe[0] >= 0);
+    assert(g_signalPipe[1] >= 0);
+    close(g_signalPipe[0]);
+    close(g_signalPipe[1]);
+    g_signalPipe[0] = -1;
+    g_signalPipe[1] = -1;
+}
+
+extern "C" void SystemNative_RegisterForCtrl(CtrlCallback callback)
+{
+    assert(callback != nullptr);
+    assert(g_ctrlCallback == nullptr);
+    g_ctrlCallback = callback;
+}
+
+extern "C" void SystemNative_UnregisterForCtrl()
+{
+    assert(g_ctrlCallback != nullptr);
+    g_ctrlCallback = nullptr;
+}
+
+extern "C" void SystemNative_RegisterForSigChld(SigChldCallback callback)
+{
+    assert(callback != nullptr);
+    assert(g_sigChldCallback == nullptr);
+    g_sigChldCallback = callback;
+}
+
+static bool InstallSignalHandler(int sig, bool overwriteIgnored = true)
+{
+    int rv;
+    struct sigaction* orig = OrigActionFor(sig);
+
+    if (!overwriteIgnored)
+    {
+        rv = sigaction(sig, NULL, orig);
+        assert(rv == 0);
+        if (reinterpret_cast<void*>(orig->sa_sigaction) == reinterpret_cast<void*>(SIG_IGN))
+        {
+            return true;
+        }
+    }
+
+    struct sigaction newAction;
+    memset(&newAction, 0, sizeof(struct sigaction));
+    newAction.sa_flags = SA_RESTART | SA_SIGINFO;
+    sigemptyset(&newAction.sa_mask);
+    newAction.sa_sigaction = &SignalHandler;
+
+    rv = sigaction(sig, &newAction, orig);
+    assert(rv == 0);
+
+    return true;
+}
+
+static bool InitializeSignalHandling()
+{
+    // Create a pipe we'll use to communicate with our worker
+    // thread.  We can't do anything interesting in the signal handler,
+    // so we instead send a message to another thread that'll do
+    // the handling work.
+    if (SystemNative_Pipe(g_signalPipe, PAL_O_CLOEXEC) != 0)
+    {
+        return false;
+    }
+    assert(g_signalPipe[0] >= 0);
+    assert(g_signalPipe[1] >= 0);
+
+    // Create a small object to pass the read end of the pipe to the worker.
+    int* readFdPtr = reinterpret_cast<int*>(malloc(sizeof(int)));
+    if (readFdPtr == nullptr)
+    {
+        CloseSignalHandlingPipe();
+        errno = ENOMEM;
+        return false;
+    }
+    *readFdPtr = g_signalPipe[0];
+
+    // The pipe is created.  Create the worker thread.
+    pthread_t handlerThread;
+    if (pthread_create(&handlerThread, nullptr, SignalHandlerLoop, readFdPtr) != 0)
+    {
+        int err = errno;
+        free(readFdPtr);
+        CloseSignalHandlingPipe();
+        errno = err;
+        return false;
+    }
+
+    // Finally, register our signal handlers
+    InstallSignalHandler(SIGINT , /* overwriteIgnored */ false);
+    InstallSignalHandler(SIGQUIT, /* overwriteIgnored */ false);
+    InstallSignalHandler(SIGCONT);
+    InstallSignalHandler(SIGCHLD);
+
+    return true;
+}
+
+extern "C" int32_t SystemNative_InitializeSignalHandling()
+{
+    static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
+    static bool initialized = false;
```

Comment:
Mono team is in the process of rewriting the System.Native PAL from .cpp to .c so that it can be shared with Mono (https://github.com/dotnet/corefx/pull/25032#issuecomment-345048294). It would be nice for more significant rewrites and additions like this file to be in .c so that they do not need to be rewritten.